src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,936 @@
+
+        :para other: The BlkDiagMatrix to compare with self.
+        :return: Returns True if no error is raised.
+        """
+
+        if not isinstance(other, BlkDiagMatrix):
+            raise NotImplementedError(
+                "Currently BlkDiagMatrix only interfaces "
+                "with its own instances, got {}".format(repr(other)))
+
+        if len(self) != len(other):
+            raise RuntimeError('Number of blocks {} {} are not equal.'.format(
+                len(self), len(other)))
+
+        self.check_size_compatible(other)
+        self.check_dtype_compatible(other)
+
+        return True
+
+    @property
+    def isfinite(self):
+        """
+        Check if all blocks in diag matrix are finite.
+
+        Calls numpy.isfinite for every entry in self.  This has the effect of
+        checking values are not += `inf` or `nan`s.
+
+        :return: Bool.
+        """
+
+        # Developers' Note Matlab port checked 0 == blk*0 for a single blk.
+        #   However, the only use case I could think of would be for the
+        #     entire matrix, so that is implemented here. Easily changed.
+        for blk in self:
+            if not np.all(np.isfinite(blk)):
+                return False
+        else:
+            return True
+
+    def add(self, other, inplace=False):
+        """
+        Define the elementwise addition of BlkDiagMatrix instance.
+
+        :param other: The rhs BlkDiagMatrix instance.
+        :param inplace: Boolean, when set to True change values in place,
+        otherwise return a new instance (default).
+        :return:  BlkDiagMatrix instance with elementwise sum equal
+        to self + other.
+        """
+
+        if self._is_scalar_type(other):
+            return self.__scalar_add(other)
+
+        self.check_compatible(other)
+
+        if inplace:
+            for i in range(self.nblocks):
+                self[i] += other[i]
+
+            C = self
+        else:
+            C = BlkDiagMatrix(self.partition, dtype=self.dtype)
+
+            for i in range(self.nblocks):
+                C[i] = self[i] + other[i]
+
+        return C
+
+    def __add__(self, other):
+        """
+        Operator overloading for addition.
+        """
+
+        return self.add(other)
+
+    def __iadd__(self, other):
+        """
+        Operator overloading for in-place addition.
+        """
+
+        if self._is_scalar_type(other):
+            return self.__scalar_add(other, inplace=True)
+
+        return self.add(other, inplace=True)
+
+    def __radd__(self, other):
+        """
+        Convenience function for elementwise scalar addition.
+        """
+
+        # Note, the case of BlkDiagMatrix_L + BlkDiagMatrix_R would be
+        #   evaluated as L.add(R), so this is only for other
+        #   Object + BlkDiagMatrix situations, namely scalars,
+        # where addition commutes
+        return self.__scalar_add(other)
```

Comment:
Not simpler to go through the standard `add` method instead of treating special cases here?